stackslib/src/net/inv/nakamoto.rs

```diff
@@ -233,3 +242,569 @@ impl InvGenerator {
+                return Err(NetError::ConnectionBroken);
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum NakamotoInvState {
+    GetNakamotoInvBegin,
+    GetNakamotoInvFinish,
+    Done,
+}
+
+/// Nakamoto inventory state machine
+pub struct NakamotoInvStateMachine<NC: NeighborComms> {
+    /// Communications links
+    pub(crate) comms: NC,
+    /// Nakamoto inventories we have
+    pub(crate) inventories: HashMap<NeighborAddress, NakamotoTenureInv>,
+    /// Reward cycle consensus hashes
+    reward_cycle_consensus_hashes: BTreeMap<u64, ConsensusHash>,
+}
+
+impl<NC: NeighborComms> NakamotoInvStateMachine<NC> {
+    pub fn new(comms: NC) -> Self {
+        Self {
+            comms,
+            inventories: HashMap::new(),
+            reward_cycle_consensus_hashes: BTreeMap::new(),
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.comms.reset();
+    }
+
+    /// Remove state for a particular neighbor
+    pub fn del_peer(&mut self, peer: &NeighborAddress) {
+        self.inventories.remove(peer);
+    }
+
+    /// Highest reward cycle learned
+    pub fn highest_reward_cycle(&self) -> u64 {
+        let mut highest_rc = 0;
+        for (_, inv) in self.inventories.iter() {
+            highest_rc = inv.highest_reward_cycle().max(highest_rc);
+        }
+        highest_rc
+    }
+
+    /// Get the consensus hash for the first sortition in the given reward cycle
+    fn load_consensus_hash_for_reward_cycle(
+        sortdb: &SortitionDB,
+        reward_cycle: u64,
+    ) -> Result<Option<ConsensusHash>, NetError> {
+        let reward_cycle_start_height = sortdb
+            .pox_constants
+            .reward_cycle_to_block_height(sortdb.first_block_height, reward_cycle);
+        let sn = SortitionDB::get_canonical_burn_chain_tip(sortdb.conn())?;
+        let ih = sortdb.index_handle(&sn.sortition_id);
+        let Some(rc_start_sn) = ih.get_block_snapshot_by_height(reward_cycle_start_height)? else {
+            return Ok(None);
+        };
+        Ok(Some(rc_start_sn.consensus_hash))
+    }
+
+    /// Populate the reward_cycle_consensus_hash mapping.  Idempotent.
+    /// Returns the current reward cycle.
+    fn update_reward_cycle_consensus_hashes(
+        &mut self,
+        sortdb: &SortitionDB,
+    ) -> Result<u64, NetError> {
+        let highest_rc =
+            if let Some((highest_rc, _)) = self.reward_cycle_consensus_hashes.last_key_value() {
+                *highest_rc
+            } else {
+                0
+            };
+
+        let sn = SortitionDB::get_canonical_burn_chain_tip(sortdb.conn())?;
+        let tip_rc = sortdb
+            .pox_constants
+            .block_height_to_reward_cycle(sortdb.first_block_height, sn.block_height)
+            .expect("FATAL: snapshot occurred before system start");
+
+        for rc in highest_rc..=tip_rc {
+            if self.reward_cycle_consensus_hashes.contains_key(&rc) {
+                continue;
+            }
```

Comment:
```suggestion
        for rc in (highest_rc+1)..=tip_rc {
```

Reply:
Sure

Reply:
Actually, this is necessary -- tests fail with out it, since we need to populate the consensus hash for `highest_rc` on the first pass.

Reply:
Is that because `highest_rc = 0` could mean either that the 0th reward cycle is the highest one processed or that no reward cycle has been processed?